### PR TITLE
HCP Telemetry Feature

### PR DESCRIPTION
**Pull request by Achooo**
### Description
This is the final PR for the merge of the HCP Telemetry feature branch (`feature/hcp-telemetry`) into `main`
A feature branch was used, where the code was thoroughly reviewed in multiple PRs, as seen by the commit history, in order:
1. #16800
2. #17100
3. #17128
4. #17159
5. #17162
6. #17184
7. #17484
8. #17455 
9. #17496 
10. Last commit for changelog

#### Feature background 
RFC: [CSL-267] Export Server Metrics to HCP

The new HCP Observability feature requires sending Consul Server metrics to HCP's new Telemetry Gateway. 
The idea is to create a new OpenTelemetry Go Metrics sink that will be initialized automatically within the fanout sink when a Consul server is registered with the HCP management plane.

minirfc_optionA drawio-10

### PR Checklist

* [ ] updated test coverage
* [ ] external facing docs updated
* [ ] appropriate backport labels added
* [ ] not a security concern

### CODE REVIEW

**Comment by Achooo** [comment]
@loshz Just opening this PR in preparation for the final merge. 
Blocked on #17455

File: .changelog/17460.txt [hunk 1]
```diff
@@ -0,0 +1,3 @@
+```release-note:feature
```
Comment (Achooo) [inline .changelog/17460.txt:1]: Any thoughts on this changelog writeup?

File: agent/hcp/client/client.go [hunk 2]
```diff
@@ -20,15 +22,34 @@ import (
...
 
+// metricsGatewayPath is the default path for metrics export request on the Telemetry Gateway.
+const metricsGatewayPath = "/v1/metrics"
+
 // Client interface exposes HCP operations that can be invoked by Consul
 //
 //go:generate mockery --name Client --with-expecter --inpackage
 type Client interface {
 	FetchBootstrap(ctx context.Context) (*BootstrapConfig, error)
+	FetchTelemetryConfig(ctx context.Context) (*TelemetryConfig, error)
```
Comment (chapmanc) [inline agent/hcp/client/client.go:33]: Note for the future. Try not to grow interfaces as much as possible. The larger the interface the weaker the abstraction. It also stops you from making composable implementations and forces you to build a mock for FetchBootstrap, when you might only want to test FetcyTelemetryConfig.

File: agent/hcp/client/client.go [hunk 3]
```diff
@@ -233,3 +280,32 @@ func (c *hcpClient) DiscoverServers(ctx context.Context) ([]string, error) {
...
+	if override := t.MetricsConfig.Endpoint; override != "" {
+		endpoint = override
+	}
+
+	if endpoint == "" {
+		return "", false
+	}
+
+	// The endpoint from Telemetry Gateway is a domain without scheme, and without the metrics path, so they must be added.
+	return endpoint + metricsGatewayPath, true
```
Comment (chapmanc) [inline agent/hcp/client/client.go:297]: Are these just strings? Ideally we use a strong type like url.

**Review by chapmanc — APPROVED** [review]
LGTM

File: agent/hcp/client/client.go (hunk 2 above)
Comment (Achooo) [reply agent/hcp/client/client.go:33]: Notes! I was trying to use the same pattern, but agreed.

File: .changelog/17460.txt (hunk 1 above)
Comment (loshz) [reply .changelog/17460.txt:1]: Maybe: 
```
hcp: Add new metrics sink to collect, aggregate and export server metrics to HCP in OTEL format.
```

**Review by loshz — APPROVED** [review]
LGTM :+1: 

Really great work on this @Achooo! This is a fantastic start and will be a great addition and starting point for observability moving forward.

File: .changelog/17460.txt (hunk 1 above)
Comment (Achooo) [reply .changelog/17460.txt:1]: Done! Thanks :D

**Pull request merged**